File: OpenEdXMobile/src/main/java/org/edx/mobile/model/api/EnrolledCoursesResponse.java
package org.edx.mobile.model.api;

import android.text.TextUtils;

import com.google.gson.annotations.SerializedName;

import org.edx.mobile.interfaces.SectionItemInterface;
import org.edx.mobile.model.course.EnrollmentMode;

import java.util.ArrayList;

@SuppressWarnings("serial")
public class EnrolledCoursesResponse implements SectionItemInterface {

    @SerializedName("audit_access_expires")
    private String auditAccessExpires;

    @SerializedName("created")
    private String created;

    @SerializedName("mode")
    private String mode;

    @SerializedName("is_active")
    private boolean isActive;

    @SerializedName("course")
    private CourseEntry course;

    private boolean isDiscussionBlackedOut = false;

    @SerializedName("certificate")
    private CertificateModel certificate;

    @SerializedName("course_modes")
    private ArrayList<CourseMode> courseModes;

    // derived fields (doesn't come in server response)
    public int videoCount;
    public long size;

    public String getAuditAccessExpires() {
        return auditAccessExpires;
    }

    public void setAuditAccessExpires(String auditAccessExpires) {
        this.auditAccessExpires = auditAccessExpires;
    }

    public String getCreated() {
        return created;
    }

    public void setCreated(String created) {
        this.created = created;
    }

    public String getMode() {
        return mode;
    }

    public void setMode(String mode) {
        this.mode = mode;
    }

    public boolean isActive() {
        return isActive;
    }

    public void setActive(boolean active) {
        this.isActive = active;
    }

    public CourseEntry getCourse() {
        return course;
    }

    public void setCourse(CourseEntry course) {
        this.course = course;
    }

    public String getCourseId() {
        return course.getId();
    }

    @Override
    public boolean isChapter() {
        return false;
    }

    @Override
    public boolean isSection() {
        return false;
    }

    @Override
    public String toString() {
        return course.getName();
    }

    @Override
    public boolean isCourse() {
        return true;
    }

    @Override
    public boolean isVideo() {
        return false;
    }

    @Override
    public boolean isDownload() {
        return false;
    }

    public String getVideoCountReadable() {
        return String.format("%d %s", videoCount, (videoCount == 1 ? "Video" : "Videos"));
    }

    public String getCertificateURL() {
        return this.certificate == null ? null : this.certificate.certificateURL;
    }

    public boolean isCertificateEarned() {
        return this.certificate != null && !TextUtils.isEmpty(this.certificate.certificateURL);

    }

    public boolean isDiscussionBlackedOut() {
        return isDiscussionBlackedOut;
    }

    public void setDiscussionBlackedOut(boolean discussionBlackedOut) {
        isDiscussionBlackedOut = discussionBlackedOut;
    }

    public String getCourseSku() {
        if (courseModes == null || courseModes.size() == 0) {
            return null;
        } else {
            for (CourseMode courseMode : courseModes) {
                if (EnrollmentMode.VERIFIED.name().equalsIgnoreCase(courseMode.getSlug())) {
                    return TextUtils.isEmpty(courseMode.getAndroidSku()) ? null : courseMode.getAndroidSku();
                }
            }
        }
        return null;
    }
}

File: OpenEdXMobile/src/main/java/org/edx/mobile/model/api/EnrolledCoursesResponse.kt
package org.edx.mobile.model.api

import androidx.recyclerview.widget.DiffUtil
import com.google.gson.annotations.SerializedName
import org.edx.mobile.interfaces.SectionItemInterface
import org.edx.mobile.model.course.EnrollmentMode
import org.edx.mobile.util.DateUtil
import java.util.Date

data class EnrolledCoursesResponse(

    @SerializedName("mode")
    var mode: String,

    @SerializedName("audit_access_expires")
    val auditAccessExpires: String?,

    @SerializedName("is_active")
    val isActive: Boolean = false,

    @SerializedName("course")
    val course: CourseEntry,

    @SerializedName("certificate")
    private val certificate: CertificateModel?,

    @SerializedName("course_modes")
    private val courseModes: List<CourseMode>?,
) : SectionItemInterface {

    var isDiscussionBlackedOut: Boolean = false

    val courseId: String
        get() = course.id

    val certificateURL: String?
        get() = certificate?.certificateURL

    val isCertificateEarned: Boolean
        get() = certificateURL.isNullOrEmpty().not()

    val courseSku: String?
        get() = courseModes?.firstOrNull { item ->
            EnrollmentMode.VERIFIED.name.equals(item.slug, ignoreCase = true)
        }?.androidSku.takeUnless { it.isNullOrEmpty() }

    val isAuditMode: Boolean
        get() = EnrollmentMode.AUDIT.toString().equals(mode, ignoreCase = true)

    val isAuditAccessExpired: Boolean
        get() = auditAccessExpires.isNullOrEmpty().not() &&
                Date().after(DateUtil.convertToDate(auditAccessExpires))

    val isUpgradeable: Boolean
        get() = isAuditMode &&
                course.isStarted &&
                course.isUpgradeDeadlinePassed.not() &&
                courseModes?.find {
                    EnrollmentMode.VERIFIED.toString().equals(it.slug, ignoreCase = true)
                } != null

    override fun isChapter(): Boolean {
        return false
    }

    override fun isSection(): Boolean {
        return false
    }

    override fun toString(): String {
        return course.name
    }

    override fun isCourse(): Boolean {
        return true
    }

    override fun isVideo(): Boolean {
        return false
    }

    override fun isDownload(): Boolean {
        return false
    }
}

/**
 * The callback for calculating the difference between two non-null items in a list.
 */
object EnrolledCoursesComparator : DiffUtil.ItemCallback<EnrolledCoursesResponse>() {
    /**
     * To check whether two objects represent the same item
     */
    override fun areItemsTheSame(
        oldItem: EnrolledCoursesResponse,
        newItem: EnrolledCoursesResponse,
    ): Boolean {
        return oldItem == newItem
    }

    /**
     * To check whether two items have the same data. With a RecyclerView.Adapter, we should return
     * whether the items' visual representations are the same.
     */
    override fun areContentsTheSame(
        oldItem: EnrolledCoursesResponse,
        newItem: EnrolledCoursesResponse,
    ): Boolean {
        return oldItem.courseId == newItem.courseId
    }
}
